### PR TITLE
Plot data resource

### PR DESCRIPTION
**Pull request by thomasballinger**
PR to track progress (or agree to forgo features)

To do:
* [x] flowtype
* [x] test this in a notebook

image

Create new issues for:

* better icons for scatter and table views
* work making plotly plots rendering async and faster by using imperative interface

### CODE REVIEW

**Comment by rgbkrk** [comment]
yayyyyyyy

**Comment by rgbkrk** [comment]
We can probably forgo making better icons, I'll get to working on that one over the next month or so.

**Comment by rgbkrk** [comment]
> work making plotly plots rendering async and faster by using imperative interface

We can leave that one up to @theengineear later on, I think it's ok to be less performant and we can fix it up later. The Plotly transform itself could probably use the same fixes, so I think it's ok to say that it's out of scope for this PR.

**Comment by thomasballinger** [comment]
@theengineear how's the chart api work? Trying this on a newer computer, making the chart transitions async and/or more efficient feels much better - so it might be ok to do a v1 without it?

File: packages/octicons/src/index.js [hunk 1]
```diff
@@ -98,6 +98,43 @@ export const LinkExternalOcticon = (props: any) => (
...
+);
+
+export const TelescopeOcticon = (props: any) => (
+  <SVGWrapper width={14} height={16} viewBox="0 0 14 16" outerProps={props}>
+    <path
+      fillRule="evenodd"
+      d="M8 9l3 6h-1l-2-4v5H7v-6l-2 5H4l2-5 2-1zM7 0H6v1h1V0zM5 3H4v1h1V3zM2 1H1v1h1V1zM.63 9a.52.52 0 0 0-.16.67l.55.92c.13.23.41.31.64.2l1.39-.66-1.16-2-1.27.86.01.01zm7.89-5.39l-5.8 3.95L3.95 9.7l6.33-3.03-1.77-3.06h.01zm4.22 1.28l-1.47-2.52a.51.51 0 0 0-.72-.17l-1.2.83 1.84 3.2 1.33-.64c.27-.13.36-.44.22-.7z"
+    />
+  </SVGWrapper>
+);
```
Comment (rgbkrk) [inline packages/octicons/src/index.js:136]: Nice! Even if we don't end up using these octicons for this data component, we'll have them ready and available for more uses. 😄

File: packages/transform-dataresource/package.json [hunk 2]
```diff
@@ -32,6 +32,8 @@
     "lodash": "^4.17.4",
     "moment": "^2.18.1",
     "react-virtualized": "^9.7.5",
-    "tv4": "^1.3.0"
+    "tv4": "^1.3.0",
```
Comment (rgbkrk) [inline packages/transform-dataresource/package.json:36]: I didn't realize we had this in here before -- we should probably use just one of the two json schema validators in this repo. You don't have to change it here though, we'll just want/need to address it sometime later to make our footprint a little smaller.

File: packages/transform-dataresource/src/index.js [hunk 3]
```diff
@@ -1,7 +1,13 @@
 /* @flow */
 import React from "react";
 import VirtualizedGrid from "./virtualized-grid";
-// import VirtualizedTable from './virtualized-table';
+import PlotlyTransform from "@nteract/transform-plotly";
+import {
+  GraphOcticon as BarGraphOcticon,
+  PulseOcticon as LineGraphOcticon,
```
Comment (rgbkrk) [inline packages/transform-dataresource/src/index.js:7]: Sneaky.

File: packages/transform-dataresource/src/index.js [hunk 4]
```diff
@@ -10,24 +16,161 @@ type Props = {
   height?: number
 };
 
-class DataResourceTransform extends React.Component<Props> {
+type State = {
+  showGrid: boolean,
+  showPlot: boolean
```
Comment (rgbkrk) [inline packages/transform-dataresource/src/index.js]: Looks like `State` should now be:

```
type State = {
  view: "line" | "bar" | "scatter" | "grid";
}
```

Yes, the literals, not the constants. Flow knows that the properties of `viewTypes` are not immutable properties.

File: packages/transform-dataresource/src/index.js [hunk 5]
```diff
@@ -10,24 +16,161 @@ type Props = {
...
+  height
+}) => {
+  return (
+    <VirtualizedGrid
+      data={data}
+      schema={schema}
+      theme={theme}
+      expanded={expanded}
+      height={height}
+      // style={{ marginRight: "10px" }}
```
Comment (rgbkrk) [inline packages/transform-dataresource/src/index.js:38]: 😄  Since we have the buttons now 😅

File: packages/transform-dataresource/src/index.js [hunk 6]
```diff
@@ -10,24 +16,161 @@ type Props = {
...
+
+type IconButtonProps = {
+  message: string,
+  onClick: () => void,
+  children?: React.Node
+};
+
+export class IconButton extends React.Component<IconButtonProps> {
+  render() {
+    const { message, onClick, children } = this.props;
```
Comment (rgbkrk) [inline packages/transform-dataresource/src/index.js:221]: Ooooh this is such a nice and easy pattern to use, sweet.

**Review by rgbkrk — APPROVED** [review]

**Comment by holdenk** [comment]
If we’re going to do push downs (since we don’t have callbacks right now), would it make sense to allow the backend to specify a push down which has already been run (eg buckited)?

**Comment by rgbkrk** [comment]
> would it make sense to allow the backend to specify a push down which has already been run (eg buckited)

What would the specified push down look like?

**Comment by rgbkrk** [comment]
⚠️ I rebased your branch ⚠️ 

Good news though!

If you do a nice fresh install at the base with:

```
lerna clean
npm install
```

Running:

```
npm run app:showcase
```

Will be wayyyyy faster, tests are cleaner, as is Travis. Happy hacking!

**Comment by rgbkrk** [comment]
Also, if `lerna clean` doesn't work well, feel free to do the `git clean -xfd` thing.

This comes by way of https://github.com/nteract/nteract/pull/2285

**Comment by rgbkrk** [comment]
One more commit and a notebook demo coming. 😉

**Comment by rgbkrk** [comment]
screen shot 2018-01-01 at 7 13 11 pm

Certainly _displays_ and can be toggled properly in the notebook, not the most useful thing though. It seems that indices should either be switched out or form the x-axis in most cases.

**Comment by rgbkrk** [comment]
screen shot 2018-01-01 at 7 52 52 pm

What do you all think of these introductory messages?

**Comment by rgbkrk** [comment]
I incorporated the `metadata.sampled` proposal from @holdenk here. It's not much of a message, but it's something. 😄

**Comment by rgbkrk** [comment]
Rebased to match up well with `master`.

**Comment by theengineear** [comment]
Rebased again to fix a merge conflict with master.

**Comment by theengineear** [comment]
@rgbkrk had to rebase again for merge conflicts on master. You suggested that it might be a good idea to just merge this in since we keep running into issues with merge conflicts. I agree 😄

**Comment by theengineear** [comment]
>work making plotly plots rendering async and faster by using imperative interface

(from the top comment)

I have some stashed changes that attempt to do this on my machine. I'd like to handle this in a separate iteration though. Perhaps we can make a new issue and just assign me to it after we merge this in?

**Comment by rgbkrk** [comment]
> Perhaps we can make a new issue and just assign me to it after we merge this in?

That works for me.

**Comment by rgbkrk** [comment]
I'm going to see about fixing tests and go ahead with merging this in. I'll then create the follow on issues.

**Comment by rgbkrk** [comment]
Alright, this passes. On to making issues!

**Pull request merged**